### PR TITLE
feat: `avm/ptn/ai-platform/baseline` New module

### PR DESCRIPTION
**Pull request by cecheta**
## Description

This is v0.1.0 of the new `avm/ptn/ai-platform/baseline` pattern: https://github.com/Azure/Azure-Verified-Modules/issues/1150

* Creates AI hub workspace, with accompanying resources
* Allows creation of compute resources

Upcoming in later versions:
* Create VNet
* Create Azure bastion + VM
* Create private endpoints from VNet to resources
* Create connections in AI hub

Note that there appears to be a bug when trying to create an AI project via ARM when the managed VNet is enabled, therefore it is not currently created by the pattern. `Managed network cannot be disabled once enabled.`

## Pipeline Reference

| Pipeline |
| -------- |
| [avm.ptn.ai-platform.baseline](https://github.com/cecheta/bicep-registry-modules/actions/workflows/avm.ptn.ai-platform.baseline.yml) |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [x] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+  'disabled'
+  'enabled'
+])
+@description('Optional. Whether the trust policy is enabled for the container registry.')
+param containerRegistryTrustPolicyStatus string = 'enabled'
+
+@description('Optional. Provide \'true\' to enable Key Vault\'s purge protection feature.')
+param keyVaultEnablePurgeProtection bool = true
+
+@description('Optional. Indicates whether the storage account permits requests to be authorized with the account access key via Shared Key. If false, then all requests, including shared access signatures, must be authorized with Azure Active Directory (Azure AD). The default value is null, which is equivalent to true.')
```
Comment (ross-p-smith) [inline avm/ptn/ai-platform/baseline/main.bicep:30]:
```suggestion
@description('Optional. Indicates whether the storage account permits requests to be authorized with the account access key via Shared Key. If false, then all requests, including shared access signatures, must be authorized with Microsoft Entra ID. The default value is null, which is equivalent to true.')
```

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+@minLength(3)
+param name string
+
+@description('Optional. Location for all Resources.')
+param location string = resourceGroup().location
+
+@description('Optional. Resource tags.')
+param tags object?
+
+@description('Optional. Enable/Disable usage telemetry for module.')
```
Comment (ross-p-smith) [inline avm/ptn/ai-platform/baseline/main.bicep:17]:
```suggestion
@description('Optional. Enable/Disable telemetry for module.')
```

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+@description('Optional. Whether the trust policy is enabled for the container registry.')
+param containerRegistryTrustPolicyStatus string = 'enabled'
+
+@description('Optional. Provide \'true\' to enable Key Vault\'s purge protection feature.')
+param keyVaultEnablePurgeProtection bool = true
+
+@description('Optional. Indicates whether the storage account permits requests to be authorized with the account access key via Shared Key. If false, then all requests, including shared access signatures, must be authorized with Azure Active Directory (Azure AD). The default value is null, which is equivalent to true.')
+param storageAccountAllowSharedKeyAccess bool = false
+
+@sys.description('Optional. Computes to create and attach to the workspace hub.')
```
Comment (ross-p-smith) [inline avm/ptn/ai-platform/baseline/main.bicep:33]: nit: computes - just sounds weird to me as a plural?!

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    networkAcls: {
+      defaultAction: 'Deny'
+      bypass: 'AzureServices'
+    }
+    publicNetworkAccess: 'Disabled'
+    enablePurgeProtection: keyVaultEnablePurgeProtection
+    roleAssignments: [
+      {
+        principalId: managedIdentity.properties.principalId
+        roleDefinitionIdOrName: 'Key Vault Administrator'
```
Comment (ross-p-smith) [inline avm/ptn/ai-platform/baseline/main.bicep:100]: Is there something smaller like `Key Vault Secrets Officer` ?

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    name: 'appi-${name}'
+    location: location
+    kind: 'web'
+    enableTelemetry: enableTelemetry
+    workspaceResourceId: logAnalyticsWorkspace.id
+    tags: tags
+  }
+}
+
+resource resourceGroup_roleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
```
Comment (ross-p-smith) [inline avm/ptn/ai-platform/baseline/main.bicep:173]: Why does the managed identity need to be contributor on the resource group (just out of interest?)

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 2 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:17]: This phrasing is pretty standard across the whole repo

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 3 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:33]: Perhaps, I just copied it from the module

But "computes" is used in documentation

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 1 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep]: Updated

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 5 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:173]: Honestly, I'm not sure

I started with just the permissions [here](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-identity-based-service-authentication?view=azureml-api-2&tabs=cli#workspace) but then ran into issues, so in the end I just went with the same as what you would get if you used system assigned managed identity.

There's also an official template [here](https://github.com/Azure/azure-quickstart-templates/tree/master/quickstarts/microsoft.machinelearningservices/machine-learning-dependencies-role-assignment) that you can deploy if you're using user assigned managed identity, and it has Contributor on the resource group

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 4 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:100]: https://github.com/Azure/bicep-registry-modules/pull/2694#discussion_r1672444855

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 5 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:173]: Just had a look now, I can see that the identity needs to create CNAME records in the private DNS zone, not sure if there's anything else?

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+  name: 'id-${name}'
+  location: location
+  tags: tags
+}
+
+resource logAnalyticsWorkspace 'Microsoft.OperationalInsights/workspaces@2023-09-01' = {
+  name: 'log-${name}'
+  location: location
+  tags: tags
+}
```
Comment (AlexanderSehr) [inline avm/ptn/ai-platform/baseline/main.bicep:79]: I take it you try to keep the number of parameters low, but I bet you'll find that user's will ask you to be able to customize this to their own naming convention in no time ;) Even though a WAF-aligned naming is definitely a good start.
Anyhow, I'd recommend to allow specifying it and use e.g. `'id-${name}` as the default for that parameter. 
The same would be true for pretty much every resource name in this module 😏 

Thoughts?

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    portRanges: string
+
+    @sys.description('Required. The protocol to allow. Provide an asterisk(*) to allow any protocol.')
+    protocol: 'TCP' | 'UDP' | 'ICMP' | '*'
+
+    @sys.description('Required. Which ports will be allow traffic by this rule. Provide an asterisk(*) to allow any port.')
+    serviceTag: string
+  }
+
+  @sys.description('')
```
Comment (AlexanderSehr) [inline avm/ptn/ai-platform/baseline/main.bicep:323]: Please note that a couple properties don't have a description. Please check.

File: avm/ptn/ai-platform/baseline/tests/e2e/max/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    location: resourceLocation
+  }
+}
+
+// General resources
+// =================
+resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
+  name: resourceGroupName
+  location: resourceLocation
+}
```
Comment (AlexanderSehr) [inline avm/ptn/ai-platform/baseline/tests/e2e/max/main.test.bicep:44]:
```suggestion
// General resources
// =================
resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: resourceGroupName
  location: resourceLocation
}

module nestedDependencies 'dependencies.bicep' = {
  scope: resourceGroup
  name: '${uniqueString(deployment().name, resourceLocation)}-nestedDependencies'
  params: {
    storageAccountName: 'dep${namePrefix}st${serviceShort}'
    location: resourceLocation
  }
}
```
Would suggest to flip the two to reflect the order of dependencies in the order of the file :)

File: avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+param baseTime string = utcNow('u')
+
+@description('Optional. A token to inject into the name of each resource.')
+param namePrefix string = '#_namePrefix_#'
+
+// ============ //
+// Dependencies //
+// ============ //
+
+module nestedDependencies 'dependencies.bicep' = {
```
Comment (AlexanderSehr) [inline avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/main.test.bicep:30]: Same comment here as in the max test regarding the order. Just makes it a bit easier to understand

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 5 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:173]: Do you think I should scope 'Contributor' to just the resources we're creating? Then, if we later run into issues we can add the missing permissions?

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 7 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:323]: Sorry, copied from the other PR before it was updated 🙈

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 5 above)
Comment (ross-p-smith) [reply avm/ptn/ai-platform/baseline/main.bicep:173]: I would follow the least privilege approach - Using Contributor is not great IMHO

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 3 above)
Comment (ross-p-smith) [reply avm/ptn/ai-platform/baseline/main.bicep:33]: another verb becomes a noun - what is happening with our language . . .

Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:33]: 😂😂

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 4 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep:99]: I think we should leave this one

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 5 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep]: I've taken Contributor off the RG and put it on the resources, let's see how it goes

File: avm/ptn/ai-platform/baseline/main.bicep (hunk 7 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/main.bicep]: Added

File: avm/ptn/ai-platform/baseline/tests/e2e/max/main.test.bicep (hunk 8 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/tests/e2e/max/main.test.bicep]: Done

File: avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/main.test.bicep (hunk 9 above)
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/main.test.bicep:37]: Done

**Comment by cecheta** [comment]
Ready for review again

**Review by Nepomuceno — CHANGES_REQUESTED** [review]
I think it would be good to make it closed by default including the storage account to be not externally accessible by defualt so they should be locked down including no internet access so this would also require you to create a vnet to put those in. The idea it is to use AVM modules instead of direct resources where possible.

File: avm/ptn/ai-platform/baseline/tests/e2e/max/dependencies.bicep [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+param storageAccountName string
+
+resource storageAccount 'Microsoft.Storage/storageAccounts@2021-09-01' = {
+  name: storageAccountName
+  location: location
+  sku: {
+    name: 'Standard_LRS'
+  }
+  kind: 'StorageV2'
+}
```
[the remark this answers is not in the document]
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/tests/e2e/max/dependencies.bicep:14]: I think it is fine to use these simple storage accounts here as they are just used for testing and destroyed straight after

The storage account that is used within the patten is using the AVM module, and I've now added SKU to be configurable

File: avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/dependencies.bicep [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+param storageAccountName string
+
+resource storageAccount 'Microsoft.Storage/storageAccounts@2021-09-01' = {
+  name: storageAccountName
+  location: location
+  sku: {
+    name: 'Standard_LRS'
+  }
+  kind: 'StorageV2'
+}
```
[the remark this answers is not in the document]
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/dependencies.bicep:14]: Also only used for testing

**Comment by cecheta** [comment]
> I think it would be good to make it closed by default including the storage account to be not externally accessible by defualt so they should be locked down including no internet access so this would also require you to create a vnet to put those in. The idea it is to use AVM modules instead of direct resources where possible.

The storage account is currently private by default and is created using an AVM module

I intend to add a VNet in a subsequent PR, for now AI studio can reach the storage account through its own managed VNet, where private endpoints are created automatically

**Review by Nepomuceno — APPROVED** [review]
LGTM

**Review by ross-p-smith — APPROVED** [review]
Great work Chinedum!

File: avm/ptn/ai-platform/baseline/main.bicep [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+output applicationInsightsName string = applicationInsights.outputs.name
+
+@description('The application ID of the application insights component.')
+output applicationInsightsApplicationId string = applicationInsights.outputs.applicationId
+
+@description('The instrumentation key of the application insights component.')
+output applicationInsightsInstrumentationKey string = applicationInsights.outputs.instrumentationKey
+
+@description('The connection string of the application insights component.')
+output applicationInsightsConnectionString string = applicationInsights.outputs.connectionString
```
Comment (AlexanderSehr) [inline avm/ptn/ai-platform/baseline/main.bicep:253]: Couldn't either of those be considered 'confidential'?

Comment (AlexanderSehr) [reply avm/ptn/ai-platform/baseline/main.bicep:253]: I guess the answer is no 😄 https://learn.microsoft.com/en-us/azure/azure-monitor/app/sdk-connection-string?tabs=dotnet5#is-the-connection-string-a-secret

File: avm/ptn/ai-platform/baseline/tests/e2e/max/dependencies.bicep (hunk 10 above)
[the remark this answers is not in the document]
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/tests/e2e/max/dependencies.bicep:14]: @Nepomuceno is happy to resolve (but doesn't seem to have permission)

File: avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/dependencies.bicep (hunk 11 above)
[the remark this answers is not in the document]
Comment (cecheta) [reply avm/ptn/ai-platform/baseline/tests/e2e/waf-aligned/dependencies.bicep:14]: @Nepomuceno is happy to resolve (but doesn't seem to have permission)

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**